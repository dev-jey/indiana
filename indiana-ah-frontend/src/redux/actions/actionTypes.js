export const REGISTER_WITH_EMAIL_REQUEST = 'REGISTER_WITH_EMAIL_REQUEST';
export const REGISTER_WITH_EMAIL_FAILURE = 'REGISTER_WITH_EMAIL_FAILURE';
export const REGISTER_WITH_EMAIL_SUCCESS = 'REGISTER_WITH_EMAIL_SUCCESS';
export const VERIFY_USER_REQUEST = 'VERIFY_USER_REQUEST';
export const VERIFY_USER_SUCCESS = 'VERIFY_USER_SUCCESS';
export const VERIFY_USER_FAILURE = 'VERIFY_USER_FAILURE';
export const REGISTER_WITH_SM = 'REGISTER_WITH_SM';
export const GET_ALL_ARTICLES = 'GET_ALL_ARTICLES';
export const NO_ARTICLES = 'NO_ARTICLES';
export const GET_ALL_ARTICLES_LOADING = 'GET_ALL_ARTICLES_LOADING';
export const GET_ALL_ARTICLES_ERROR = 'GET_ALL_ARTICLES_ERROR';
export const LOGIN_WITH_EMAIL_REQUEST = 'LOGIN_WITH_EMAIL_REQUEST';
export const LOGIN_WITH_EMAIL_FAILURE = 'LOGIN_WITH_EMAIL_FAILURE';
export const LOGIN_WITH_EMAIL_SUCCESS = 'LOGIN_WITH_EMAIL_SUCCESS';
export const NETWORK_FAILURE = 'NETWORK_FAILURE';
export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const SIGN_OUT_USER = 'SIGN_OUT_USER';
export const GET_ALL_BOOKMARKS = 'GET_ALL_BOOKMARKS';
export const GET_ALL_BOOKMARKS_LOADING = 'GET_ALL_BOOKMARKS_LOADING';
export const GET_ALL_BOOKMARKS_FAILURE = 'GET_ALL_BOOKMARKS_FAILURE';
export const ADD_BOOKMARK = 'ADD_BOOKMARK';
export const REMOVE_BOOKMARK = 'REMOVE_BOOKMARK';
export const CREATE_ARTICLE = 'CREATE_ARTICLE';
export const CREATE_ARTICLE_FAILURE = 'CREATE_ARTICLE_FAILURE';
export const CREATE_ARTICLE_LOADING = 'CREATE_ARTICLE_LOADING';
export const GET_ALL_USER_ARTICLES_REQUEST = 'GET_ALL_USER_ARTICLES_REQUEST';
export const GET_ALL_USER_ARTICLES_SUCCESS = 'GET_ALL_USER_ARTICLES_SUCCESS';
export const GET_ALL_USER_ARTICLES_FAILURE = 'GET_ALL_USER_ARTICLES_FAILURE';
export const GET_SINGLE_ARTICLE_SUCCESS = 'GET_SINGLE_ARTICLE_SUCCESS';
export const GET_SINGLE_ARTICLE_LOADING = 'GET_SINGLE_ARTICLE_LOADING';
export const SEND_EMAIL_LOADING = 'SEND_EMAIL_LOADING';
export const SEND_EMAIL_SUCCESS = 'SEND_EMAIL_SUCCESS';
export const SEND_EMAIL_FAILURE = 'SEND_EMAIL_FAILURE';
export const LIKE_ARTICLE = 'LIKE_ARTICLE';
export const DISLIKE_ARTICLE = 'DISLIKE_ARTICLE';
export const DELETE_ARTICLE_LOADING = 'DELETE_ATRICLE_LOADING';
export const DELETE_ARTICLE_SUCCESS = 'DELETE_ARTICLE_SUCCESS';
export const DELETE_ARTICLE_FAILURE = 'DELETE_ARTICLE_FAILURE';
export const GET_USER_PROFILE_REQUEST = 'GET_USER_PROFILE_REQUEST';
export const GET_USER_PROFILE_SUCCESS = 'GET_USER_PROFILE_SUCCESS';
export const GET_USER_PROFILE_ERROR = 'GET_USER_PROFILE_ERROR';
export const UPDATE_USER_PROFILE_SUCCESS = 'UPDATE_USER_PROFILE_SUCCESS';
export const UPDATE_USER_PROFILE_ERROR = 'UPDATE_USER_PROFILE_ERROR';
export const UPDATE_USER_PROFILE_REQUEST = 'UPDATE_USER_PROFILE_REQUEST';
export const UPDATE_USER_PASSWORD_SUCCESS = 'UPDATE_USER_PASSWORD_SUCCESS';
export const UPDATE_USER_PASSWORD_ERROR = 'UPDATE_USER_PASSWORD_ERROR';
export const UPDATE_USER_PASSWORD_REQUEST = 'UPDATE_USER_PASSWORD_REQUEST';
export const UPDATE_IN_APP_NOTIFICATION_SUCCESS = 'UPDATE_IN_APP_NOTIFICATION_SUCCESS';
export const UPDATE_IN_APP_NOTIFICATION_ERROR = 'UPDATE_IN_APP_NOTIFICATION_ERROR';
export const UPDATE_IN_APP_NOTIFICATION_REQUEST = 'UPDATE_IN_APP_NOTIFICATION_REQUEST';
export const UPDATE_EMAIL_NOTIFICATION_SUCCESS = 'UPDATE_EMAIL_NOTIFICATION_SUCCESS';
export const UPDATE_EMAIL_NOTIFICATION_ERROR = 'UPDATE_EMAIL_NOTIFICATION_ERROR';
export const UPDATE_EMAIL_NOTIFICATION_REQUEST = 'UPDATE_EMAIL_NOTIFICATION_REQUEST';
export const UPDATE_PROFILE_PIC_SUCCESS = 'UPDATE_PROFILE_PIC_SUCCESS';
export const UPDATE_PROFILE_PIC_ERROR = 'UPDATE_PROFILE_PIC_ERROR';
export const UPDATE_PROFILE_PIC_REQUEST = 'UPDATE_PROFILE_PIC_REQUEST';
export const AUTHENTICATE_USER = 'AUTHENTICATE_USER';
export const ADD_COMMENT = 'ADD_COMMENT';
export const GET_ALL_ARTICLE_COMMENTS = 'GET_ALL_ARTICLE_COMMENTS';
export const COMMENTS_LOADING = 'COMMENTS_LOADING';
export const DELETE_COMMENT = 'DELETE_COMMENT';
export const SEARCH_ARTICLE_REQUEST = 'SEARCH_ARTICLE_REQUEST';
export const SEARCH_ARTICLE_SUCCESS = 'SEARCH_ARTICLE_SUCCESS';
export const SEARCH_ARTICLE_FAILURE = 'SEARCH_ARTICLE_FAILURE';
export const SEND_RESET_LINK_LOADING = 'SEND_RESET_LINK_LOADING';
export const SEND_RESET_LINK_SUCCESS = 'SEND_RESET_LINK_SUCCESS';
export const SEND_RESET_LINK_FAILURE = 'SEND_RESET_LINK_FAILURE';
export const UPDATE_PASSWORD_LOADING = 'UPDATE_PASSWORD_LOADING';
export const UPDATE_PASSWORD_SUCCESS = 'UPDATE_PASSWORD_SUCCESS';
export const UPDATE_PASSWORD_FAILURE = 'UPDATE_PASSWORD_FAILURE';
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const EDIT_COMMENTS_FAILURE = 'EDIT_COMMENTS_FAILURE';
export const LIKE_COMMENT = 'LIKE_COMMENT';
export const DISLIKE_COMMENT = 'DISLIKE_COMMENT';
export const GET_USER_STATS_REQUEST = 'GET_USER_STATS_REQUEST';
export const GET_USER_STATS_SUCCESS = 'GET_USER_STATS_SUCCESS';
export const GET_USER_STATS_ERROR = 'GET_USER_STATS_ERROR';
export const GET_COMMENT_EDIT_HISTORY = 'GET_COMMENT_EDIT_HISTORY';
export const COMMENT_EDIT_HISTORY_LOADING = 'COMMENT_EDIT_HISTORY_LOADING';

export const GET_ALL_USER_FOLLOWING_LOADING = 'GET_ALL_USER_FOLLOWING_LOADING';
export const GET_ALL_USER_FOLLOWING_FAILURE = 'GET_ALL_USER_FOLLOWING_FAILURE';
export const GET_ALL_USER_FOLLOWING_SUCCESS = 'GET_ALL_USER_FOLLOWING_SUCCESS';

export const UNFOLLOW_USER_FAILURE = 'UNFOLLOW_USER_FAILURE';
export const UNFOLLOW_USER_SUCCESS = 'UNFOLLOW_USER_SUCCESS';
export const UNFOLLOW_USER_LOADING = 'UNFOLLOW_USER_LOADING';

export const FOLLOW_USER_LOADING = 'FOLLOW_USER_LOADING';
export const FOLLOW_USER_FAILURE = 'FOLLOW_USER_FAILURE';
export const FOLLOW_USER_SUCCESS = 'FOLLOW_USER_SUCCESS';

export const GET_ALL_USER_FOLLOWERS_LOADING = 'GET_ALL_USER_FOLLOWERS_LOADING';
export const GET_ALL_USER_FOLLOWERS_FAILURE = 'GET_ALL_USER_FOLLOWERS_FAILURE';
export const GET_ALL_USER_FOLLOWERS_SUCCESS = 'GET_ALL_USER_FOLLOWERS_SUCCESS';
