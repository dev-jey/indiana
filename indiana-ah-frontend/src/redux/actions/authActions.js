import React from 'react';
import { toast } from 'react-toastify';
import {
  REGISTER_WITH_EMAIL_REQUEST,
  REGISTER_WITH_EMAIL_FAILURE,
  REGISTER_WITH_EMAIL_SUCCESS,
  LOGIN_WITH_EMAIL_REQUEST,
  LOGIN_WITH_EMAIL_FAILURE,
  LOGIN_WITH_EMAIL_SUCCESS,
  VERIFY_USER_REQUEST,
  VERIFY_USER_FAILURE,
  VERIFY_USER_SUCCESS,
  NETWORK_FAILURE,
  SET_CURRENT_USER,
  SIGN_OUT_USER,
  REGISTER_WITH_SM,
  SEND_EMAIL_FAILURE,
  SEND_EMAIL_LOADING,
  SEND_EMAIL_SUCCESS,
  SEND_RESET_LINK_LOADING,
  SEND_RESET_LINK_SUCCESS,
  SEND_RESET_LINK_FAILURE,
  UPDATE_PASSWORD_LOADING,
  UPDATE_PASSWORD_FAILURE,
  UPDATE_PASSWORD_SUCCESS
} from './actionTypes';
import { sendHttpRequest, validateToken } from '../../utils/index';

export const registerWithEmail = (data, { closeModal, history }) => async (dispatch) => {
  dispatch({ type: REGISTER_WITH_EMAIL_REQUEST });
  try {
    const { message, token } = await sendHttpRequest('/register', 'POST', data);
    localStorage.setItem('token', token);
    const user = validateToken(token);
    dispatch({ type: REGISTER_WITH_EMAIL_SUCCESS });
    dispatch({ type: SET_CURRENT_USER, user });
    closeModal();
    history.push('/');
    toast.success(<div>{message}</div>);
  } catch ({ response }) {
    if (!response) {
      dispatch({ type: NETWORK_FAILURE, payload: 'Connection error.' });
      toast.error(<div>Network Error. Check your connection</div>);
      return;
    }

    dispatch({
      type: REGISTER_WITH_EMAIL_FAILURE,
      payload: response.data.message
    });
  }
};

export const loginWithEmail = (data, { closeModal }) => async (dispatch) => {
  dispatch({ type: LOGIN_WITH_EMAIL_REQUEST });
  try {
    const { message, token } = await sendHttpRequest('/login', 'POST', data);
    localStorage.clear();
    localStorage.setItem('token', token);
    const user = validateToken(token);
    dispatch({ type: SET_CURRENT_USER, user });
    dispatch({ type: LOGIN_WITH_EMAIL_SUCCESS });
    closeModal();
    toast.success(<div>{message}</div>);
  } catch ({ response }) {
    dispatch({
      type: LOGIN_WITH_EMAIL_FAILURE,
      payload: response.data.message
    });
  }
};

export const verifyUser = (query, { history }) => async (dispatch) => {
  dispatch({ type: VERIFY_USER_REQUEST });
  try {
    const { token } = await sendHttpRequest(`/users/verify?query=${query}`, 'PATCH');
    localStorage.setItem('token', token);
    dispatch({ type: VERIFY_USER_SUCCESS });
    toast.success(<div>Email verification successful</div>);
    history.push('/');
  } catch ({ response }) {
    if (!response) {
      dispatch({ type: NETWORK_FAILURE, payload: 'Connection error.' });
      toast.error(<div>Network Error. Check your connection</div>);
      history.push('/');
      return;
    }

    dispatch({ type: VERIFY_USER_FAILURE, payload: response.data.message });
    toast.error(<div>Email verification failed</div>);
    history.push('/');
  }
};

export const signOutUser = () => (dispatch) => {
  window.localStorage.clear();
  dispatch({ type: SIGN_OUT_USER });
};

export const loginWithSocialMedia = token => (dispatch) => {
  const user = validateToken(token);
  if (user) {
    localStorage.setItem('token', token);
    dispatch({ type: SET_CURRENT_USER, user });
    dispatch({ type: REGISTER_WITH_SM });
  } else {
    toast.error(<div>We cannot log you in by this time. Please try again later.</div>);
  }
};

export const sendUserEmail = ({ history }) => async (dispatch) => {
  dispatch({ type: SEND_EMAIL_LOADING });
  try {
    const { message, token } = await sendHttpRequest('/verify/email', 'POST');
    localStorage.setItem('token', token);
    dispatch({ type: SEND_EMAIL_SUCCESS });
    toast.success(<div>{message}</div>);
    history.push('/');
  } catch (res) {
    dispatch({ type: SEND_EMAIL_FAILURE });
  }
};

export const sendResetLink = (data, { closeModal }) => async (dispatch) => {
  dispatch({ type: SEND_RESET_LINK_LOADING });
  try {
    const { message } = await sendHttpRequest('/users/begin-password-reset',
      'POST', data);
    localStorage.clear();
    dispatch({ type: SEND_RESET_LINK_SUCCESS, payload: 'Success' });
    toast.success(<div>{message}</div>);
    closeModal();
  } catch ({ response }) {
    dispatch({ type: SEND_RESET_LINK_FAILURE, payload: response.data.message });
    toast.error(<div>{response.data.message}</div>);
  }
};

export const resetPassword = (data, token, { history }) => async (dispatch) => {
  dispatch({ type: UPDATE_PASSWORD_LOADING });
  try {
    const { message } = await sendHttpRequest(`/users/reset-password?query=${token}`,
      'PATCH', data);
    dispatch({ type: UPDATE_PASSWORD_SUCCESS, payload: 'Success' });
    toast.success(<div>{message}</div>);
    history.push('/');
  } catch ({ response }) {
    dispatch({ type: UPDATE_PASSWORD_FAILURE, payload: response.data.message });
    toast.error(<div>{response.data.message}</div>);
  }
};
