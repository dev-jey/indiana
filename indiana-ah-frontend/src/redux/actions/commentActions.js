import { toast } from 'react-toastify';
import {
  ADD_COMMENT,
  GET_ALL_ARTICLE_COMMENTS,
  DELETE_COMMENT,
  COMMENTS_LOADING,
  EDIT_COMMENT,
  EDIT_COMMENTS_FAILURE
} from './actionTypes';
import { sendHttpRequest, addUserReaction } from '../../utils/index';

export const addComment = (articleSlug, commentData) => async (dispatch, getState) => {
  const user = getState().user.userData;
  const commenter = {
    name: user.name,
    username: user.username,
    imageUrl: user.imageUrl
  };
  dispatch({ type: COMMENTS_LOADING });
  try {
    const response = await sendHttpRequest(
      `/articles/${articleSlug}/comments`,
      'POST',
      commentData
    );
    const commentDetails = response.data;
    commentDetails.commenter = commenter;
    commentDetails.likes = 0;
    commentDetails.dislikes = 0;
    return dispatch({ type: ADD_COMMENT, payload: commentDetails });
  } catch ({ response }) {
    switch (response.status) {
      case 404:
      case 401:
        return toast.error(response.data.message);
      default:
        return toast.error(`Cannot comment on this article at the moment.
          Please try again later`);
    }
  }
};

export const getArticleComments = articleSlug => async (dispatch, getState) => {
  const { id } = getState().user.userData;
  try {
    const response = await sendHttpRequest(`/articles/${articleSlug}/comments`, 'GET');
    const comments = response.comments.map((comment) => {
      const likes = comment.CommentReactions.filter(
        reaction => reaction.reactionType === 'like'
      );
      const dislikes = comment.CommentReactions.filter(
        reaction => reaction.reactionType === 'dislike'
      );
      comment.likes = likes.length;
      comment.dislikes = dislikes.length;
      addUserReaction(comment, 'CommentReactions', id);
      return comment;
    });
    return dispatch({ type: GET_ALL_ARTICLE_COMMENTS, payload: comments });
  } catch ({ response }) {
    switch (response.status) {
      case 404:
      case 401:
        return toast.error(response.data.message);
      default:
        return toast.error(`Cannot fetch the comments on this article at the moment. 
          Please try again later`);
    }
  }
};

export const deleteComment = id => async (dispatch) => {
  dispatch({ type: COMMENTS_LOADING });
  try {
    const response = await sendHttpRequest(`/comments/${id}`, 'DELETE', {
      commentId: id
    });
    toast.success(response.message);
    return dispatch({ type: DELETE_COMMENT, id });
  } catch (error) {
    switch (error.response.status) {
      case 404:
      case 401:
        return toast.error(error.response.data.message);
      default:
        return toast.error(`Cannot delete this comment at the moment.
          Please try again later`);
    }
  }
};

export const editComment = (id, commentData) => async (dispatch, getState) => {
  const user = getState().user.userData;
  const commenter = {
    name: user.name,
    username: user.username,
    imageUrl: user.imageUrl
  };
  const commentBody = commentData;
  dispatch({ type: COMMENTS_LOADING });
  try {
    const response = await sendHttpRequest(`/articles/comments/${id}/`, 'PUT', {
      commentBody
    });
    const commentDetails = response.comment;
    commentDetails.commenter = commenter;
    dispatch({ type: EDIT_COMMENT, payload: commentDetails });
    return response.message;
  } catch ({ response }) {
    dispatch({ type: EDIT_COMMENTS_FAILURE, payload: response });
  }
};
