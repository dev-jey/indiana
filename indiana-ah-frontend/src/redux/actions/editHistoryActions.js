import { toast } from 'react-toastify';
import {
  GET_COMMENT_EDIT_HISTORY, COMMENT_EDIT_HISTORY_LOADING
} from './actionTypes';
import { sendHttpRequest } from '../../utils/index';

export default commentId => async (dispatch) => {
  dispatch({ type: COMMENT_EDIT_HISTORY_LOADING });
  try {
    const response = await sendHttpRequest(`/articles/comments/${commentId}/history`, 'GET');
    return dispatch({ type: GET_COMMENT_EDIT_HISTORY, payload: response.commentEditHistory });
  } catch ({ response }) {
    switch (response.status) {
      case 404:
      case 401:
        return toast.error(response.data.message);
      default:
        return toast.error(`Cannot fetch the edited comments at the moment. 
          Please try again later`);
    }
  }
};
