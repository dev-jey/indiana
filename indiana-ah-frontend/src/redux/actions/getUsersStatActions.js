import React from 'react';
import { toast } from 'react-toastify';
import {
  GET_USER_STATS_REQUEST,
  GET_USER_STATS_SUCCESS,
  GET_USER_STATS_ERROR
} from './actionTypes';
import { sendHttpRequest } from '../../utils/index';

const getUserStat = () => async (dispatch) => {
  dispatch({ type: GET_USER_STATS_REQUEST });
  try {
    const {
      message: { statistics }
    } = await sendHttpRequest('/users/statistics', 'GET');
    dispatch({ type: GET_USER_STATS_SUCCESS, payload: [statistics] });
  } catch (error) {
    dispatch({
      type: GET_USER_STATS_ERROR,
      payload: 'There was an error fetching your statistics'
    });
    toast.error(<div>There was an error fetching your statistics</div>);
  }
};
export default getUserStat;
