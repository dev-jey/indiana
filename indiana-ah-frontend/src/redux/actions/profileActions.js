import React from 'react';
import { toast } from 'react-toastify';
import {
  GET_USER_PROFILE_ERROR,
  GET_USER_PROFILE_REQUEST,
  GET_USER_PROFILE_SUCCESS,
  UPDATE_USER_PROFILE_ERROR,
  UPDATE_USER_PROFILE_REQUEST,
  UPDATE_USER_PROFILE_SUCCESS,
  UPDATE_USER_PASSWORD_SUCCESS,
  UPDATE_USER_PASSWORD_ERROR,
  UPDATE_USER_PASSWORD_REQUEST,
  UPDATE_IN_APP_NOTIFICATION_SUCCESS,
  UPDATE_IN_APP_NOTIFICATION_ERROR,
  UPDATE_IN_APP_NOTIFICATION_REQUEST,
  UPDATE_EMAIL_NOTIFICATION_SUCCESS,
  UPDATE_EMAIL_NOTIFICATION_ERROR,
  UPDATE_EMAIL_NOTIFICATION_REQUEST,
  UPDATE_PROFILE_PIC_SUCCESS,
  UPDATE_PROFILE_PIC_ERROR,
  UPDATE_PROFILE_PIC_REQUEST
} from './actionTypes';
import { sendHttpRequest } from '../../utils/index';

const subscribedToInApp = 'You have successfully subscribed to in app notifications';
const subscribedToEmail = 'You have successfully subscribed to our email notifications';

export const getUserProfile = () => async (dispatch, getState) => {
  dispatch({ type: GET_USER_PROFILE_REQUEST });
  const {
    user: {
      userData: { username }
    }
  } = getState();
  try {
    const { profile } = await sendHttpRequest(`/profiles/${username}`, 'GET');
    dispatch({ type: GET_USER_PROFILE_SUCCESS, payload: profile });
  } catch (error) {
    dispatch({
      type: GET_USER_PROFILE_ERROR,
      payload: 'There was an error fetching your profile'
    });
    toast.error(<div>There was an error fetching your profile</div>);
  }
};
export const updateUserProfile = data => async (dispatch, getState) => {
  dispatch({ type: UPDATE_USER_PROFILE_REQUEST });
  const {
    user: {
      userData: { username }
    }
  } = getState();
  try {
    const { profile } = await sendHttpRequest(
      `/profiles/${username}/update`,
      'PATCH',
      data
    );
    dispatch({ type: UPDATE_USER_PROFILE_SUCCESS, payload: profile });
    toast.success(<div>updated profile successfully</div>);
  } catch ({
    response: {
      data: { message }
    }
  }) {
    dispatch({
      type: UPDATE_USER_PROFILE_ERROR,
      payload: message
    });
    toast.error(<div>There was an error updating your profile</div>);
  }
};

export const updatePassword = data => async (dispatch, getState) => {
  dispatch({ type: UPDATE_USER_PASSWORD_REQUEST });
  const {
    user: {
      userData: { username }
    }
  } = getState();
  try {
    await sendHttpRequest(`/profiles/${username}/password`, 'PATCH', data);
    dispatch({ type: UPDATE_USER_PASSWORD_SUCCESS });
    toast.success(<div>updated password successfully</div>);
  } catch ({
    response: {
      data: { message }
    }
  }) {
    dispatch({
      type: UPDATE_USER_PASSWORD_ERROR,
      payload: message
    });
    toast.error(<div>There was an error updating your password</div>);
  }
};
export const updateInAppNotification = data => async (dispatch) => {
  dispatch({ type: UPDATE_IN_APP_NOTIFICATION_REQUEST });
  try {
    const { message } = await sendHttpRequest('/notifications/inApp', 'PATCH', data);
    if (message === subscribedToInApp) {
      dispatch({
        type: UPDATE_IN_APP_NOTIFICATION_SUCCESS,
        payload: { inAppNotification: true }
      });
      return toast.success(<div>{message}</div>);
    }
    dispatch({
      type: UPDATE_IN_APP_NOTIFICATION_SUCCESS,
      payload: { inAppNotification: false }
    });
    return toast.warn(<div>{message}</div>);
  } catch ({
    response: {
      data: { message }
    }
  }) {
    dispatch({
      type: UPDATE_IN_APP_NOTIFICATION_ERROR,
      payload: message
    });
    toast.error(<div>There was an error subscribing to the InApp notification</div>);
  }
};

export const updateEmailNotification = data => async (dispatch) => {
  dispatch({ type: UPDATE_EMAIL_NOTIFICATION_REQUEST });
  try {
    const { message } = await sendHttpRequest('/notifications/email', 'PATCH', data);
    if (message === subscribedToEmail) {
      dispatch({
        type: UPDATE_EMAIL_NOTIFICATION_SUCCESS,
        payload: { subscribed: true }
      });
      return toast.success(<div>{message}</div>);
    }
    dispatch({
      type: UPDATE_EMAIL_NOTIFICATION_SUCCESS,
      payload: { subscribed: false }
    });
    return toast.warn(<div>{message}</div>);
  } catch ({
    response: {
      data: { message }
    }
  }) {
    dispatch({
      type: UPDATE_EMAIL_NOTIFICATION_ERROR,
      payload: message
    });
    toast.error(<div>There was an error subscribing to email notification</div>);
  }
};
export const updateProfilePic = data => async (dispatch, getState) => {
  dispatch({ type: UPDATE_PROFILE_PIC_REQUEST });
  const {
    user: {
      userData: { username }
    }
  } = getState();
  try {
    const formdata = new FormData();
    formdata.append('image', data);
    const {
      data: { imageUrl },
      message
    } = await sendHttpRequest(`/profiles/${username}/image`, 'PATCH', formdata);
    dispatch({
      type: UPDATE_PROFILE_PIC_SUCCESS,
      payload: { imageUrl }
    });
    return toast.success(<div>{message}</div>);
  } catch ({
    response: {
      data: { message }
    }
  }) {
    dispatch({
      type: UPDATE_PROFILE_PIC_ERROR,
      payload: message
    });
    toast.error(<div>{message}</div>);
  }
};
