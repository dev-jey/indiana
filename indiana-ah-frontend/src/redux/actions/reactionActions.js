import { sendHttpRequest } from '../../utils';
import {
  LIKE_ARTICLE,
  DISLIKE_ARTICLE,
  LIKE_COMMENT,
  DISLIKE_COMMENT
} from './actionTypes';

export const reactToArticle = (slug, reactionType) => async (dispatch) => {
  dispatch({
    type: reactionType === 'like' ? LIKE_ARTICLE : DISLIKE_ARTICLE
  });
  try {
    await sendHttpRequest(`/articles/${slug}/reaction`, 'POST', {
      reactionType
    });
  } catch (error) {
    return error;
  }
};

export const reactToComment = (id, reactionType) => async (dispatch) => {
  dispatch({
    type: reactionType === 'like' ? LIKE_COMMENT : DISLIKE_COMMENT,
    payload: id
  });
  try {
    await sendHttpRequest('/comments/reaction', 'POST', {
      reactionType,
      commentId: id
    });
  } catch (error) {
    return error;
  }
};
