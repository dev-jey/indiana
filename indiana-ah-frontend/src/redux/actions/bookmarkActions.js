import { toast } from 'react-toastify';
import { ADD_BOOKMARK, REMOVE_BOOKMARK } from './actionTypes';
import { sendHttpRequest } from '../../utils/index';

const addBookmark = id => async (dispatch, getState) => {
  const Article = getState().singleArticle.article;
  try {
    const response = await sendHttpRequest(`articles/${id}/bookmark`, 'post');
    if (response.bookmark) {
      const bookmarkDetails = response.bookmark;
      bookmarkDetails.Article = Article;
      toast.success(response.message);
      return dispatch({ type: ADD_BOOKMARK, payload: bookmarkDetails });
    }
    toast.success(response.message);
    return dispatch({ type: REMOVE_BOOKMARK, id });
  } catch (error) {
    switch (error.response.status) {
      case 404:
      case 401:
        return toast.error(error.response.data.message);
      default:
        return toast.error(`Cannot bookmark this article at the moment. 
          Please try again later`);
    }
  }
};
export default addBookmark;
