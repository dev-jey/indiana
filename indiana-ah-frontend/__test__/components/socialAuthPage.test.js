import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import jwt from 'jsonwebtoken';
import ConnectedSocialAuthPage from '../../src/components/SocialAuthPage.jsx';
import reducer from '../../src/redux/reducers/authReducer';
import { REGISTER_WITH_SM, SET_CURRENT_USER } from '../../src/redux/actions/actionTypes';
import { loginWithSocialMedia } from '../../src/redux/actions/authActions';

const mockStore = configureStore([thunk]);
const store = mockStore();

describe('Social auth page component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <ConnectedSocialAuthPage />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('social auth reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual({
      isLoading: false,
      isAuthenticated: false,
      isVerified: false,
      error: '',
      sendResetLink: '',
      updatePassword: ''
    });
  });
  it('should setup reducer to login with social media', () => {
    const successAction = {
      type: REGISTER_WITH_SM
    };
    expect(reducer({}, successAction)).toEqual({
      isAuthenticated: true
    });
  });
});

describe('Auth action creators test', () => {
  const token = jwt.sign({ name: 'omenkish' }, 'yeeeeeeeeeeee', { expiresIn: '24hrs' });
  beforeEach(() => {
    store.clearActions();
  });

  it('should create the REGISTER_WITH_SM and SET_CURRENT_USER actions', () => {
    store.dispatch(loginWithSocialMedia(token));
    expect(store.getActions()).toEqual([
      {
        type: SET_CURRENT_USER,
        user: {
          name: 'omenkish'
        }
      },
      { type: REGISTER_WITH_SM }
    ]);
  });

  it('should fail if token is not provided', () => {
    store.dispatch(loginWithSocialMedia());
    expect(store.getActions()).toEqual([]);
  });
});
