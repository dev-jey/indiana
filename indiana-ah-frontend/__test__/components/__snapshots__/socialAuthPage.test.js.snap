// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social auth page component renders correctly 1`] = `
<div>
  <p>
    We are verifying your account.....
  </p>
</div>
`;
